site/app/models/CalendarInfo.php

```diff
@@ -0,0 +1,131 @@
+     * @param GradeableList $gradeable_list container of gradeables in the system
+     * @return CalendarInfo
+     */
+    public static function loadGradeableCalendarInfo(Core $core, GradeableList $gradeable_list): CalendarInfo {
+        $info = new CalendarInfo($core);
+
+        // get the gradeables from the GradeableList and group them by section
+        $gradeable_list_sections = [
+            GradeableList::OPEN => $gradeable_list->getOpenGradeables(),
+            GradeableList::GRADING => $gradeable_list->getGradingGradeables(),
+            GradeableList::CLOSED => $gradeable_list->getClosedGradeables(),
+            GradeableList::GRADED => $gradeable_list->getGradedGradeables(),
+        ];
+
+        foreach ($gradeable_list_sections as $section => $gradeables) {
+            /** @var int $section */
+            /** @var array<Gradeable> $gradeables */
+            $curr_section["title"] = NavigationView::gradeableSections[$section]["title"];
+            $curr_section["subtitle"] = NavigationView::gradeableSections[$section]["subtitle"];
+            $curr_section["section_id"] = NavigationView::gradeableSections[$section]["section_id"];
+            $curr_section['gradeables'] = [];
+
+            // Iterate over the Gradeable objects in current section and summarize data
+            foreach ($gradeables as $id => $gradeable) {
+                /** @var Gradeable $gradeable */
+                [$semester, $course_title, $gradeable_id] = explode("||", $id);
+                $currGradeable = [
+                    'gradeable_id' => $gradeable_id,
+                    'title' => $gradeable->getTitle(),
+                    'semester' => $semester,
+                    'course' => $course_title,
+                    'url' => $info->core->buildUrl(['courses', $semester, $course_title, 'gradeable', $gradeable->getId()]),
+                    'submission' => ($gradeable->getType() === GradeableType::ELECTRONIC_FILE) ? $gradeable->getSubmissionDueDate() : '',
+                    'status' => (string) $section,
+                    'grading_open' => $gradeable->getGradeStartDate(),
+                    'grading_due' => $gradeable->getGradeDueDate()
+                ];
+
+                // Put gradeables in current section by their id which consists of semester, course title and gradeable id
+                $curr_section['gradeables'][$id] = $currGradeable;
+
+                // Put gradeables in by-date maps according to section (close/open)
+                $dueDate = ($gradeable->getType() === GradeableType::ELECTRONIC_FILE) ? $gradeable->getSubmissionDueDate()->format('Y-m-d') : '';
+                $info->items_by_date[$dueDate][] = $currGradeable;
+            }
+
+            // Put data of current section into the by-section map
+            $info->items_by_sections[] = $curr_section;
+        }
+
+        $info->empty_message = "There are currently no assignments posted.  Please check back later.";
```

Comment:
Do we need to define this string within this factory method, or can we just assign it to `private $empty_message = ''` on line 49 above? Or probably better yet just put this string within the twig template and get rid of this variable altogether.

Reply:
This string may be changed in different factory methods. Currently CalendarInfo handles Gradeables, but in future commits the calendar will be used for course materials (Course Calendar feature) and this string will be contents related with course materials